stumpy/scrump.py

```diff
@@ -103,67 +109,111 @@ def _compute_PI(
     for i in indices[start:stop]:
         Q = T_A[i : i + m]
         QT[:] = core._sliding_dot_product(Q, T_B)
-        # Update P[i] relative to all T[j : j + m]
         squared_distance_profile[:] = core._mass(Q, T_B, QT, μ_Q[i], σ_Q[i], M_T, Σ_T)
         squared_distance_profile[:] = np.square(squared_distance_profile)
         if excl_zone is not None:
-            zone_start = max(0, i - excl_zone)
-            zone_stop = min(l, i + excl_zone)
-            squared_distance_profile[zone_start : zone_stop + 1] = np.inf
-
-            # only for self-join
-            mask = squared_distance_profile < P_squared[thread_idx]
-            P_squared[thread_idx][mask] = squared_distance_profile[mask]
-            I[thread_idx][mask] = i
-
-        I[thread_idx, i] = np.argmin(squared_distance_profile)
-        P_squared[thread_idx, i] = squared_distance_profile[I[thread_idx, i]]
-        if P_squared[thread_idx, i] == np.inf:  # pragma: no cover
-            I[thread_idx, i] = -1
-        else:
-            j = I[thread_idx, i]
-            # Given the squared distance, work backwards and compute QT
-            QT_j = (m - P_squared[thread_idx, i] / 2.0) * (Σ_T[j] * σ_Q[i]) + (
-                m * M_T[j] * μ_Q[i]
+            core._apply_exclusion_zone(squared_distance_profile, i, excl_zone, np.inf)
+
+        # find EXACT (not approx.) value of `P_squared[thread_idx, i, 0]`
+        nn = np.argmin(squared_distance_profile)
+        core._shift_insert_at_index(
+            P_squared[thread_idx, i], 0, squared_distance_profile[nn]
+        )
+        core._shift_insert_at_index(I[thread_idx, i], 0, nn)
+
+        if P_squared[thread_idx, i, 0] == np.inf:  # pragma: no cover
+            I[thread_idx, i, 0] = -1
+            continue
+
+        j = I[thread_idx, i, 0]
+        # Given the squared distance, work backwards and compute QT
+        QT_j = (m - P_squared[thread_idx, i, 0] / 2.0) * (Σ_T[j] * σ_Q[i]) + (
+            m * M_T[j] * μ_Q[i]
+        )
+        QT_j_prime = QT_j
+        # Update Top-k of BOTH subsequences at i+g and j+g (i.e. right neighbor
```

Comment:
Can we say `the subsequence S_{i+j}`? because `T[i+g]` feels like an element rather than a subsequence.

Reply:
"Update top-k for both subsequences `S[i+g] = T[i+g:i+g+m]` and `S[j+g] = T[j+g:j+g+m]` (i.e., the right neighbors of `T[i : i+m]` and `T[j:j+m]`) by using the distance between `S[i+g]` and `S[j+g]`"

Reply:
Note that "top-k" is lower case "t". I notice that you tend to use uppercase "Top-k" so please try to avoid it except at the start of s sentence

Reply:
> "Update top-k for both subsequences `S[i+g] = T[i+g:i+g+m]` and `S[j+g] = T[j+g:j+g+m]` (i.e., the right neighbors of `T[i : i+m]` and `T[j:j+m]`) by using the distance between `S[i+g]` and `S[j+g]`"

Sounds good. I will revise it.

> Note that "top-k" is lower case "t". I notice that you tend to use uppercase "Top-k" so please try to avoid it except at the start of s sentence

Thanks for catching that! I will try to correct it in other places when I see it.